guardian/support-service-lambdas
# feat: write lambda polling SQS and creating a single contribution record in Salesforce

## What does this change?

- Implement lambda handler polling the SQS queue and creating a single contribution record in Salesforce
- Add Salesforce auth credentials secrets in AWS Secrets Manager
- Add policy statement to lambda role policy to allow `secretsmanager:GetSecretValue` of Salesforce auth credentials
- (health) Delete all files related to the obsolete project failed-national-deliveries-processor

[**Trello Card**](https://trello.com/c/JCxMwBRv/335-single-contribution-record-create-lambda-to-create-a-single-contribution-record-in-salesforce)

## Why are you doing this?

This is part of the epic [Create a single contribution record in Salesforce via support-payment-api](https://trello.com/c/qibQVaR2/295-create-a-single-contribution-record-in-salesforce-via-support-frontend-payment-api).

Previous work:

1. [Create single-contribution-salesforce-writes project with SQS queues to handle single contribution messages](https://github.com/guardian/support-service-lambdas/pull/2059)
2. [Add rule to acquisition bus to send payment-api messages to the single-contribution-salesforce-writes-queue](https://github.com/guardian/support-service-lambdas/pull/2066)
3. [Add fields (postalCode, state, email) to AcquisitionDataRow](https://github.com/guardian/support-frontend/pull/5345)

## Architecture
![Screenshot 2023-10-23 at 10 41 47](https://github.com/guardian/support-service-lambdas/assets/39066365/f0b0dc98-0feb-4dd6-a8ec-1b7f32f123b6)

## Project structure

I have attempted to create a standardised folder structure:

- scala
  - models
    - Errors
  - handlers
    - CreateSalesforceSingleContributionRecordHandler
  - services
    - awsSecretsManager
      - GetSecretValue
    - jsonDecoder
      - DecodeJson
    - salesforce
      - CreateSingleContributionRecord
      - GetAccessToken
      - GetApiDomain
- test (to be completed) 

## Design notes
  
- The lambda handler is separated from the core logic as suggested [here](https://docs.aws.amazon.com/lambda/latest/dg/best-practices.html)
- CloudWatch alarm strategy needs to be added to the diagram
- PROD Salesforce secrets to be added before deploying to production

## How to test

1. Make a single contribution from the [CODE site](https://support.code.dev-theguardian.com/uk/contribute)
2. From the Salesforce sandbox dashboard, check that the contribution record has been created [here](https://gnmtouchpoint--dev1.sandbox.lightning.force.com/lightning/r/Report/00O9E000001A5vkUAC/view)

## How can we measure success?

Single contributions made via payment-api should match single contribution records in Salesforce.

## Next steps

- Add alarm if messages appear in dead letter queue
- Add README / docs with architectural diagram (C4 diagram experimentation)
- Add dead letter queue to event bus rule to prevent messages getting lost
- Add tests

## Thank you
Thanks to @shtukas and @tomrf1 for your help on Scala!
